util/src/test/java/tech/pegasys/artemis/util/async/SafeFutureTest.java

```diff
@@ -485,6 +485,45 @@ public void fromRunnable_propagatesExceptionalResult() {
     assertExceptionallyCompletedWith(future, error);
   }
 
+  @Test
+  public void allOf_shouldAddAllSuppressedExceptions() {
+    final Throwable error1 = new RuntimeException("Nope");
+    final Throwable error2 = new RuntimeException("Oh dear");
+    final SafeFuture<Void> future1 = new SafeFuture<>();
+    final SafeFuture<Void> future2 = new SafeFuture<>();
+    final SafeFuture<Void> future3 = new SafeFuture<>();
+
+    final SafeFuture<Void> result = SafeFuture.allOf(future1, future2, future3);
+    assertThat(result).isNotDone();
+
+    future2.completeExceptionally(error2);
+    assertThat(result).isNotDone();
+
+    future3.complete(null);
+    assertThat(result).isNotDone();
+
+    future1.completeExceptionally(error1);
+
+    assertThat(result).isCompletedExceptionally();
+    assertThatThrownBy(result::join).hasSuppressedException(error2);
+    assertThatThrownBy(result::join).hasRootCause(error1);
```

Comment:
What makes the `error1` the root cause here and not the `error2`?

Reply:
Just the way `CompletableFuture` works.  I was expecting error2 as well to be honest.